### PR TITLE
Hapi 17: Fix inability to augment server.plugins type

### PR DESCRIPTION
**Pull request by jhsimms**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/hapi/blob/master/API.md#server.plugins
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by jhsimms** [comment]
@SimonSchick @lenovouser I wish I had seen this in the last PR, but ran into it this morning.  The server.plugins type is `any`.  This change lets us type the returned value.

File: types/hapi/index.d.ts
Comment (SimonSchick) [inline types/hapi/index.d.ts]: `Exposed` is kinda redundant.

**Review by SimonSchick — APPROVED** [review]
looks good.

**Comment by SimonSchick** [comment]
Actually, can you apply the same to `RouteOptions.app`?

**Comment by SimonSchick** [comment]
`RouteOptions.app` 
`ServerInjectOptions.app` should be the same interface as above.
`ServerMethodOptions.generateKey` returns `string | null` 
`ServerOptions.app`

This fixes a bunch of `any`s!

**Comment by SimonSchick** [comment]
@lenovouser I guess so, these typings are better than the previous ones for sure, we got rid of a lot of the type ambigiouties use of vague types.

**Comment by SimonSchick** [comment]
One more nit:

`request.query` is technically not any but 

```typescript
 { [key: string]: string | string[]; } | string
```

**Comment by SimonSchick** [comment]
@jhsimms sorry for the feature creep 😛

**Comment by jhsimms** [comment]
Np on the feature creep, @SimonSchick, I fixed the other things you noticed.

**Comment by lenovouser** [comment]
Aaaah, @jhsimms you ninja'd me - compiling a project fails for me with

```sh
yarn run v1.3.2
$ tsc -p .
node_modules/@types/hapi/index.d.ts:1374:20 - error TS2314: Generic type 'AnySchema<T>' requires 1 type argument(s).

1374     [key: string]: AnySchema;
                        ~~~~~~~~~


error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by lenovouser** [comment]
I am not sure yet why it throws that

**Comment by SimonSchick** [comment]
@lenovouser are you still referencing joi v6?

**Comment by lenovouser** [comment]
@SimonSchick inside my project? No. Only import is `hapi`

**Comment by SimonSchick** [comment]
For some reason it seems your project references joi v6, can you pinpoint it?

**Comment by lenovouser** [comment]
@SimonSchick I think it was some npm caching issue, I cleared the npm cache, removed `node_modules`, reinstalled and recompiled. Now it's working...

I never used / referenced `joi` in this project anyways. Not before I started updating it to v17 and not while doing it. Weird.

I love technology `¯\_(ツ)_/¯`

**Review by lenovouser — APPROVED** [review]

File: types/hapi/index.d.ts
Comment (SimonSchick) [inline types/hapi/index.d.ts]: From what I can tell these should be the same as `RouteOptionsApp `.

Comment (jhsimms) [reply types/hapi/index.d.ts]: The description says `sets the initial value of request.app, defaults to {}.`.  The type of request.app in our typings now is currently ApplicationState.  Would that be the correct type?

Comment (jhsimms) [reply types/hapi/index.d.ts]: I changed it as I have it above, as I think it is correct.  If I'm misreading, let me know and I can change it to how you suggested.

**Review by SimonSchick — APPROVED** [review]
re-approving also @andy-ms I think this is good to go.

**Comment by andy-ms** [comment]
Approved by third party. PR appears ready to merge pending review by a maintainer.

**Pull request merged**
